### PR TITLE
Irmin and modular backend support

### PR DESCRIPTION
**Pull request by MagnusS**
PR to track Irmin support and support for new vm backends (xapi + libxl).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,8 @@
-PACKAGES=-package lwt.syntax,lwt,dns.lwt,libvirt,cmdliner,ezxmlm,ipaddr,str,conduit,conduit.lwt-unix
+PACKAGES=-package lwt.syntax,lwt,dns.lwt,libvirt,cmdliner,ezxmlm,ipaddr,str,conduit,conduit.lwt-unix,xen-api-client,xen-api-client.lwt,irmin.unix
 INCLUDE=
 OPT=-linkpkg -g 
 OCAMLOPT=ocamlopt -w A-4-44
-FILES=synjitsu.mli synjitsu.ml jitsu.mli jitsu.ml main.ml
+FILES=vm_stop_mode.ml vm_state.ml backends.mli libvirt_backend.ml xapi_backend.ml irmin_backend.mli irmin_backend.ml dns_helpers.ml synjitsu.mli synjitsu.ml jitsu.mli jitsu.ml main.ml
```
Comment (samoht) [inline Makefile]: any specific reason to use a Makefile here instead of using `oasis`? I can make a patch for that if you want.

Comment (MagnusS) [inline Makefile]: No, not really - it was just convenient when the project was smaller. A patch would be very welcome :-)

File: src/backends.mli [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  val connect : string -> [ `Ok of t | `Error of error ] Lwt.t
+  (** **)
+  val lookup_vm_by_uuid : t -> uuid -> [ `Ok of vm | `Error of error ] Lwt.t
+  (** Lookup a VM by UUID *)
+  val lookup_vm_by_name : t -> string -> [ `Ok of vm | `Error of error ] Lwt.t
+  (** Lookup UUID of a VM by name *)
+
+  val get_state : t -> vm -> [ `Ok of Vm_state.t | `Error of error ] Lwt.t
+  (** Get VM state from [vm] type *)
```
Comment (samoht) [inline src/backends.mli]: you might need to add blank lines between the statements as this will confuse `ocamldoc`.

File: src/jitsu.ml [hunk 3]
```diff
@@ -16,314 +16,283 @@
...
     in
-    (* Notify Synjitsu *)
-    (match vm.mac with
-      | Some m -> 
-              (match t.synjitsu with
-              | Some s -> (
+    let notify_synjitsu () =
+      Irmin_backend.get_ip t.storage ~vm_name >>= fun r ->
+      match r with
+      | Some ip -> begin
```
Comment (samoht) [inline src/jitsu.ml]: this pattern matching is a bit awful :p 

You can either do a top-level one:

```ocaml
match r, vm_mac, t.synjitsu, with
| Some ip, Some m, Some s -> [...]
| _ -> Lwt.return_unit
```
or use a function to compose them.

**Comment by samoht** [comment]
The patches look great! I haven't looked at all the details, but I like the general shape :-)

**Comment by MagnusS** [comment]
Thanks! It's still work in progress though... but hopefully I'll have something that we can merge in a few days :-)

**Comment by MagnusS** [comment]
This PR should now be more or less complete. The new backends (xapi, libxl) still need more testing.

**Pull request merged**